### PR TITLE
add prototype image folder dataset

### PR DESCRIPTION
**Pull request by pmeier**
This is the precursor of the new style datasets. Example:

```python
from torchvision.prototype import datasets

dataset, categories = datasets.from_image_folder(...)
dataset = dataset.shuffle().decode_sample(datasets.decoder.pil, "image")

for sample in dataset:
    ...
```

cc @pmeier

### CODE REVIEW

**Comment by pmeier** [comment]
@fmassa Should we have a `prototype` label for these PRs to easily tell them apart from other stuff for the release notes?

File: torchvision/prototype/datasets/decoder.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import numpy as np
+import PIL.Image
+import torch
+from torch.utils.data import IterDataPipe, functional_datapipe
+
+__all__ = ["pil"]
+
+
+@functional_datapipe("decode_sample")
```
Comment (fmassa) [inline torchvision/prototype/datasets/decoder.py:13]: I've mentioned this to @VitalyFedyunin in the past, but I think that the fluent API should be used with care.

The reason is that things like
```python
import torch.utils.data
import torchvision.prototype

generic_datapipe.decode_sample(...)
```
works but
```python
import torch.utils.data

generic_datapipe.decode_sample(...)
```
doesn't can be very surprising to users, and it is very very hard to spot what is wrong / missing.

IMO I'm not sure this is a pattern we would like to encourage outside of PyTorch core abstractions.
Thoughts?

**Comment by pmeier** [comment]
One problem I see is that we cannot decode until after the `.shuffle()` call: it will read `buffer_size` samples ([defaults to 10000](https://github.com/pytorch/pytorch/blob/9157a2889f9e17dda1be6b0b137ab1801ba253a9/torch/utils/data/datapipes/iter/combinatorics.py#L79)) into memory before a single sample is returned to the user. 

Given that in most of the cases the user will want to shuffle the dataset, I'm wondering if we even want to provide an option for decoding in the future API. If we don't we need to figure out a good way to pass on the information which keys of the sample need to be decoded from the dataset to the user.

File: torchvision/prototype/datasets/_folder.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    for label, category in enumerate(categories):
+        category_dp = FileLister(str(root / category), recursive=True)
+        category_dp = FileLoader(category_dp)
+        category_dp = Mapper(
+            category_dp,
+            _collate,
+            fn_kwargs=dict(label=label, category=category),
+        )
+        category_dps.append(category_dp)
+    return Concater(*category_dps), categories
```
Comment (fmassa) [inline torchvision/prototype/datasets/_folder.py:44]: Doesn't this make shuffling much harder, because the elements will be sampled from the same category in order? Or does the shuffler have additional knowledge that knows how to handle those use-cases in a special way?

File: torchvision/prototype/datasets/_folder.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+def _collate(
+    data: Tuple[str, io.IOBase],
+    *,
+    label: int,
+    category: str,
+) -> Dict[str, Any]:
+    path, image_file = data
+    return dict(
+        path=path,
+        image=image_file,
```
Comment (fmassa) [inline torchvision/prototype/datasets/_folder.py:22]: nit: maybe calling this `image` might be misleading because the image hasn't been decoded yet?

**Review by fmassa — COMMENTED** [review]
That's a great start, thanks!

I have a few questions for my understanding, let me know what you think

**Comment by fmassa** [comment]
> @fmassa Should we have a prototype label for these PRs to easily tell them apart from other stuff for the release notes?

Yes, good idea, can you create one?

File: torchvision/prototype/datasets/decoder.py (hunk 1 above)
Comment (pmeier) [reply torchvision/prototype/datasets/decoder.py:13]: Agreed. Apart from the availability, this is also hard to discover. Unless you grep for `decode_sample`, you probably won't find the `decode_sample` datapipe.

**Comment by datumbox** [comment]
I'm brining up a question I raised previously concerning the name of this space. A few were proposed over time, is `prototype` the selected term across all domain libs?

**Comment by fmassa** [comment]
@datumbox torchaudio follows `prototype` as well, and torchtext had traditionally used `experimental`. IIRC we had an agreement to use `prototype` in a past discussion

File: torchvision/prototype/datasets/_folder.py (hunk 2 above)
Comment (pmeier) [reply torchvision/prototype/datasets/_folder.py:44]: > Doesn't this make shuffling much harder, because the elements will be sampled from the same category in order?

Yes, but I have no idea how to improve this. Maybe we can read one sample from each directory before we start at the top again. Or we implement a `RandomChoser` which picks one element at random from multiple datapipes.

Comment (fmassa) [reply torchvision/prototype/datasets/_folder.py:44]: Can't we just have a single `FileLister` which is recursive? This way it makes the job of a `Shuffler` much easier, as FileLister could implement a specific property that allows for perfect shuffling for those family of pipes.

@VitalyFedyunin @ejguan thoughts?

File: torchvision/prototype/datasets/_folder.py (hunk 3 above)
Comment (pmeier) [reply torchvision/prototype/datasets/_folder.py:22]: Valid question. IMO decoding should be an in-place operation and since in most cases there will be some decoding at some point, I feel like "image" is ok as key. Plus, otherwise the user would need to specify the name of the decoded feature together with the feature to be decoded, e.g.

```python
dataset.decode_sample(datasets.decoder.pil, ("image_file", "image")
```

File: torchvision/prototype/datasets/_folder.py (hunk 2 above)
Comment (pmeier) [reply torchvision/prototype/datasets/_folder.py:44]: > Can't we just have a single `FileLister` which is recursive?

We can, yes. As far as I see, that would make sorting harder, since we now need to extract the label and category from each path instead of just setting them. I'll implement it that way and benchmark.

File: torchvision/prototype/datasets/_folder.py (hunk 3 above)
Comment (fmassa) [reply torchvision/prototype/datasets/_folder.py:22]: We can revisit this in the early future, so let's not block on this.

File: torchvision/prototype/datasets/decoder.py (hunk 1 above)
Comment (fmassa) [reply torchvision/prototype/datasets/decoder.py:13]: So let's remove this from torchvision and be more verbose on our end, wdyt?

**Comment by pmeier** [comment]
Workflow now looks like:

```python
from torchvision.prototype import datasets

dataset, categories = datasets.from_image_folder(...)
decode = datasets.decoder.decode_sample(datasets.decoder.pil, "image")

dataset = dataset.shuffle().map(decode)

for sample in dataset:
    ...

```

File: torchvision/prototype/datasets/decoder.py (hunk 1 above)
Comment (pmeier) [reply torchvision/prototype/datasets/decoder.py]: See https://github.com/pytorch/vision/pull/4441#issuecomment-921851202

File: torchvision/prototype/datasets/_folder.py (hunk 2 above)
Comment (pmeier) [reply torchvision/prototype/datasets/_folder.py:44]: If we use

```python
class RandomPicker(IterDataPipe):
    def __init__(self, *datapipes):
        self.datapipes = datapipes

    def __iter__(self):
        non_exhausted = [iter(dp) for dp in self.datapipes]
        while non_exhausted:
            dp = random.choice(non_exhausted)
            try:
                yield next(dp)
            except StopIteration:
                non_exhausted.remove(dp)
```

instead of `Concater` we get a pseudo-shuffled dataset with minimal overhead.

**Comment by VitalyFedyunin** [comment]
Hi! As we likely going to have shuffle topic again, I captured notes in the doc file. Feel free to comment/feedback. https://docs.google.com/document/d/15RzQtCMl2FDtu9loZH5a-wxHDPnSpT0LRT6wMjR1p-U/edit

File: torchvision/prototype/datasets/datapipes.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+"""This is a temporary module for datapipes that are needed now, but should be absorbed by torchdata in the future."""
+
+import random
+from typing import Any
+
+from torch.utils.data import IterDataPipe
+
+
+class RandomPicker(IterDataPipe):
```
Comment (VitalyFedyunin) [inline torchvision/prototype/datasets/datapipes.py:9]: https://github.com/facebookexternal/torchdata/blob/7881a93c6f83aa5c6a129c2123a42cf23c7dddd9/torchdata/datapipes/iter/__init__.py#L29 might help

File: torchvision/prototype/datasets/_folder.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+
+def from_image_folder(
+    root: Union[str, pathlib.Path], *, pseudo_shuffle: bool = True
+) -> Tuple[IterDataPipe, List[str]]:
+    root = pathlib.Path(root).expanduser().resolve()
+    categories = sorted({item.name for item in os.scandir(root) if item.is_dir})
+    category_dps = []
+    category_dp: IterDataPipe
+    for label, category in enumerate(categories):
```
Comment (VitalyFedyunin) [inline torchvision/prototype/datasets/_folder.py:38]: This is interesting approach to avoid scanning twice to obtain list of categories.

File: torchvision/prototype/datasets/datapipes.py (hunk 4 above)
Comment (pmeier) [reply torchvision/prototype/datasets/datapipes.py:9]: I'll test when `torchdata` is public.

File: torchvision/prototype/datasets/_folder.py (hunk 5 above)
Comment (pmeier) [reply torchvision/prototype/datasets/_folder.py:38]: Interesting as in good or bad interesting?

**Comment by pmeier** [comment]
I've refactored the loading process to be able to shuffle directly after the files are enumerated. The workflow now looks like this:

```python
from torchvision.prototype import datasets

dataset = datasets.from_image_folder(root)
```

By default, the images will now be shuffled with an infinite buffer, i.e. a true random permutation. To turn this off one can pass `datasets.from_image_folder(..., shuffler=None)`.

`datasets.from_image_folder(root)` is now equivalent to the current `datasets.ImageFolder(root, transform=transforms.ToTensor())`. Testing this on Caltech256 data, leads to almost identical warm-up times and the datapipe implementation is even a little faster for completely consuming the dataset.

File: torchvision/prototype/datasets/_folder.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+from torch.utils.data import IterDataPipe
+from torch.utils.data.datapipes.iter import FileLister, Concater, FileLoader, Mapper, Shuffler
+
+from torchvision.prototype.datasets.decoder import pil
+
+
+__all__ = ["from_data_folder", "from_image_folder"]
+
+# pseudo-infinite buffer size until a true infinite buffer is supported
+INFINITE = 1_000_000_000
```
Comment (pmeier) [inline torchvision/prototype/datasets/_folder.py:18]: @VitalyFedyunin @ejguan We need the ability to specify an infinite buffer, i.e. `buffer_size=None` for every datapipe that implements a buffer.

File: torchvision/prototype/datasets/decoder.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+import numpy as np
+import PIL.Image
+import torch
+
+__all__ = ["pil"]
+
+
+def pil(file: io.IOBase, mode="RGB") -> torch.Tensor:
+    image = PIL.Image.open(file).convert(mode.upper())
+    return torch.from_numpy(np.array(image, copy=True)).permute((2, 0, 1))
```
Comment (fmassa) [inline torchvision/prototype/datasets/decoder.py:12]: FYI this could be replaced by `torchvision.transforms.functional.pil_to_tensor`

File: torchvision/prototype/datasets/decoder.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+import numpy as np
+import PIL.Image
+import torch
+
+__all__ = ["pil"]
+
+
+def pil(file: io.IOBase, mode="RGB") -> torch.Tensor:
+    image = PIL.Image.open(file).convert(mode.upper())
```
Comment (fmassa) [inline torchvision/prototype/datasets/decoder.py:11]: Do we want to close the file, or doesn't it make a difference? Maybe using a context manager here?

**Review by fmassa — APPROVED** [review]
Let's get this merged.

I have some comments that can be addressed in a follow-up PR

File: torchvision/prototype/datasets/decoder.py (hunk 8 above)
Comment (pmeier) [reply torchvision/prototype/datasets/decoder.py:11]: IIRC, we don't need to. `.convert()` does this automatically.

File: torchvision/prototype/datasets/_folder.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        data=data,
+        label=label,
+        category=category,
+    )
+
+
+def from_data_folder(
+    root: Union[str, pathlib.Path],
+    *,
+    shuffler: Optional[Callable[[IterDataPipe], IterDataPipe]] = lambda dp: Shuffler(dp, buffer_size=INFINITE),
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/prototype/datasets/_folder.py:48]: I think not shuffling by default is ok. Although everyone will probably turn it on to do anything useful with the dataset, we now also require users to turn it on manually.
